File: parser/src/test/java/com/github/fburato/highwheelmodules/bytecodeparser/ClassPathParserTest.java
package com.github.fburato.highwheelmodules.bytecodeparser;

import static org.mockito.Mockito.*;

import java.io.IOException;
import java.io.InputStream;
import java.util.Collections;
import java.util.function.Predicate;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;
import com.github.fburato.highwheelmodules.model.classpath.AccessVisitor;
import com.github.fburato.highwheelmodules.model.classpath.ClasspathRoot;
import com.github.fburato.highwheelmodules.model.bytecode.ElementName;

public class ClassPathParserTest {

    private ClassPathParser testee;

    @Mock
    private ClasspathRoot cp;

    @Mock
    private Predicate<ElementName> filter;

    @Mock
    private AccessVisitor v;

    @BeforeEach
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        this.testee = new ClassPathParser(this.filter);
    }

    @Test
    public void shouldNotParseClassesThatDoNotMatchFilter() throws IOException {
        final ElementName foo = ElementName.fromString("foo");
        when(this.cp.classNames()).thenReturn(Collections.singleton(foo));
        when(this.filter.test(foo)).thenReturn(false);
        testee.parse(cp, v);
        verify(cp, never()).getData(foo);
    }

    @Test
    public void shouldCloseClassInputStreams() throws IOException {
        final ElementName foo = ElementName.fromString("foo");
        when(this.cp.classNames()).thenReturn(Collections.singleton(foo));
        when(this.filter.test(foo)).thenReturn(true);
        final InputStream is = Mockito.mock(InputStream.class);
        when(this.cp.getData(foo)).thenReturn(is);
        doThrow(new IOException()).when(is).read();
        doThrow(new IOException()).when(is).read(any(), anyInt(), anyInt());
        try {
            this.testee.parse(cp, this.v);
        } catch (final IOException ex) {
            // expected
        }
        verify(is).close();

    }

}

File: parser/src/test/scala/com/github/fburato/highwheelmodules/bytecodeparser/ClassPathParserTest.scala
package com.github.fburato.highwheelmodules.bytecodeparser

import com.github.fburato.highwheelmodules.bytecodeparser.TryMatchers._
import com.github.fburato.highwheelmodules.model.bytecode.ElementName
import com.github.fburato.highwheelmodules.model.classpath.{AccessVisitor, ClasspathRoot}
import org.mockito.scalatest.MockitoSugar
import org.scalatest.OneInstancePerTest
import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

import java.io.{IOException, InputStream}
import scala.util.Success

class ClassPathParserTest
    extends AnyWordSpec
    with Matchers
    with MockitoSugar
    with OneInstancePerTest {

  private val classPathRoot = mock[ClasspathRoot]
  private val fooString = "foo"
  private val testee = new ClassPathParser(element => element.asInternalName == fooString)

  "parse" should {
    "not parse classes that do not match filter" in {
      val bar = ElementName.fromString("bar")
      when(classPathRoot.classNames).thenReturn(Success(Seq(bar)))

      testee.parse(classPathRoot, mock[AccessVisitor]).get

      verify(classPathRoot, never).getData(any)
    }

    "close class input streams even if they fail" in {
      val foo = ElementName.fromString(fooString)
      when(classPathRoot.classNames).thenReturn(Success(Seq(foo)))
      val is = mock[InputStream]
      when(is.available()).thenReturn(1)
      when(classPathRoot.getData(foo)).thenReturn(Success(is))
      val exception = new IOException()
      doThrow(exception).when(is).read(any, anyInt, anyInt)

      testee.parse(classPathRoot, mock[AccessVisitor]) should beFailureWith[Unit] { exception =>
        exception shouldBe theSameInstanceAs(exception)
      }

      verify(is).close()
    }
  }
}
